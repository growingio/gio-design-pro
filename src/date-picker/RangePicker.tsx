/* eslint-disable @typescript-eslint/no-shadow */
import React, { useState, useEffect } from 'react';
import Button from '@gio-design/components/es/components/button';
import List from '@gio-design/components/es/components/list';
import Dropdown from '@gio-design/components/es/components/dropdown';
import usePrefixCls from '@gio-design/components/es/utils/hooks/use-prefix-cls';
import moment, { Moment } from 'moment';
import { CalendarOutlined } from '@gio-design/icons';
import Calendar from './Calendar';
import { RangePickerProps } from './interface';

enum Mode {
  shortcut = 'shortcut',
  since = 'since',
  dynamic = 'dynamic',
  absolute = 'absolute',
}

const RangePicker: React.FC<RangePickerProps> = (props: RangePickerProps) => {
  const { prefixCls: customizePrefixCls } = props;
  const [v, setV] = useState(false);
  const [mode, setMode] = useState<Mode>(Mode.shortcut);
  const [displayTime, setDisplayTime] = useState('今天');
  // const [formatedTime, setFormatedTime] = useState(timeRange);
  const modeOptions = [
    { value: Mode.shortcut, label: '常用时间' },
    { value: Mode.since, label: '自某天以后' },
    { value: Mode.dynamic, label: '过去动态时段' },
    { value: Mode.absolute, label: '过去固定时段' },
  ];
  const shortcutOptions = [
    [
      { value: 'day:1,0', label: '今日' },
      { value: 'week:1,0', label: '本周' },
      { value: 'month:1,0', label: '本月' },
      { value: 'quarter:1,0', label: '本季度' },
      { value: 'year:1,0', label: '今年' },
      { value: 'day:8,1', label: '过去7天' },
      { value: 'day:31,1', label: '过去30天' },
      { value: 'day:181,1', label: '过去180天' },
      { value: 'hour:24,0', label: '过去24小时' },
      { value: 'since', label: '开始至今' },
    ],
    [
      { value: 'day:2,1', label: '昨日' },
      { value: 'week:2,1', label: '上周' },
      { value: 'month:2,1', label: '上月' },
      { value: 'quarter:2,1', label: '上季度' },
      { value: 'year:2,1', label: '去年' },
      { value: 'day:15,1', label: '过去14天' },
      { value: 'day:91,1', label: '过去90天' },
      { value: 'day:366,1', label: '过去365天' },
      { value: 'hour:48,0', label: '过去48小时' },
    ],
  ];

  useEffect(() => {
    if (v) {
      setMode(Mode.shortcut);
    }
  }, [v]);

  const prefixCls = usePrefixCls('date-picker', customizePrefixCls);
  const [time, setTime] = useState([moment(new Date()), moment(new Date())]);
  // const [nextTime, setNextTime] = useState([moment(new Date()), moment(new Date())]);
  const format = 'YYYY/MM/DD';
  const formatDate = (_: Moment) => _.format(format);

  const formatDisplayRange = (mode: Mode, value: Array<Moment>) => {
    if (mode === Mode.since) {
      const sinceStart = formatDate(value[0]);
      const sinceEnd = value[1].day() === moment().day() ? '今天' : '昨天';
      return `自 ${sinceStart} 至${sinceEnd}`;
    }
    if (mode === Mode.dynamic) {
      const dynamicRight = value[1].diff(moment(), 'days');
      const dynamicLeft = value[0].diff(moment(), 'days');
      return dynamicRight === 0 ? `过去${-dynamicLeft}天` : `过去${-dynamicRight}-${-dynamicLeft}天`;
    }
    if (mode === Mode.absolute) {
      return `从 ${formatDate(value[0])} 到 ${formatDate(value[1])} `;
    }
    return '';
  };

  const onListClick = (item: any) => {
    setDisplayTime(item.label);
    // setFormatedTime(item.value);
    setV(false);
  };

  const formatRange = (mode: Mode, value: Array<Moment>) => {
    if (mode === Mode.since) {
      // todo
    }
    if (mode === Mode.dynamic) {
      const dynamicRight = value[1].diff(moment(), 'days');
      const dynamicLeft = value[0].diff(moment(), 'days');
      return `day:${-dynamicLeft + 1},${-dynamicRight + 1}`;
    }
    if (mode === Mode.absolute) {
      return `abs:${value[0].valueOf()},${value[1].valueOf()}`;
    }
    return '';
  };

  const onChange = (value: Array<Moment>) => {
    setDisplayTime(formatDisplayRange(mode, value));
    formatRange(mode, value);
    value && setTime(value);
  };

  const onConfirm = () => {
    setV(false);
  };

  const onCancel = () => {
    setV(false);
  };

  const handleModeChange = (item: any) => {
    setMode(item.value);
  };

  const renderCalendar = () => {
    return (
      <div
        style={{
          boxSizing: 'border-box',
          width: '540px',
          height: '430px',
          position: 'relative',
          display: 'block',
          lineHeight: 1.5,
          marginBottom: 16,
        }}
      >
        <Calendar
          prefixCls={prefixCls}
          value={time}
          onChange={onChange}
          format={format}
          onCancel={onCancel}
          onConfirm={onConfirm}
          mode={mode}
          showFooter
        />
      </div>
    );
  };

  const renderShortcuts = () => {
    return (
      <div
        style={{
          display: 'inline-flex',
        }}
      >
        <List stateless dataSource={shortcutOptions[0]} width={128} onClick={onListClick} />
        <List stateless dataSource={shortcutOptions[1]} width={128} onClick={onListClick} />
      </div>
    );
  };

  return (
    <Dropdown
      visible={v}
      onVisibleChange={setV}
      overlay={
        <div
          style={{
            width: 'auto',
            height: 'auto',
            backgroundColor: '#FFFFFF',
            display: 'inline-flex',
            justifyContent: 'center',
          }}
        >
          <List
            wrapStyle={{ borderRight: '1px solid #E7EAF9' }}
            dataSource={modeOptions}
            width={144}
            onClick={handleModeChange}
          />
          <div>{mode === Mode.shortcut ? renderShortcuts() : renderCalendar()}</div>
        </div>
      }
      placement="bottomRight"
    >
      <Button type="secondary" icon={<CalendarOutlined />}>
        {displayTime}
      </Button>
    </Dropdown>
  );
};

export default RangePicker;
