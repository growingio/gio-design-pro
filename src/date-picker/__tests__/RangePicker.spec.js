import React from 'react';
import { mount, render } from 'enzyme';
import { noop } from 'lodash';
import RangePicker from '../DateRangePicker';

const VALUE = 'day:2,1';

const props = {
  value: VALUE,
  onChange: noop,
};

describe('DatePicker ui test', () => {
  const DatePickerInstance = () => <RangePicker {...props} />;

  it('should match snapshot', () => {
    const wrapper = render(DatePickerInstance());
    expect(wrapper).toMatchSnapshot();
  });

  it('should receive props', () => {
    expect(() => {
      const wrapper = mount(DatePickerInstance());
      wrapper.setProps({ value: 'day:2,1' });
      wrapper.setProps({ onChange: noop });
      wrapper.unmount();
    }).not.toThrow();
  });
});
