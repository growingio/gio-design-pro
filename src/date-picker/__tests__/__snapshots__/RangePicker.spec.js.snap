// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DatePicker ui test should match snapshot 1`] = `
<button
  class="gio-btn gio-btn-secondary"
  style="cursor:pointer"
  type="button"
>
  <span
    aria-label="calendar-outlined"
    class="gio-icon"
    role="img"
  >
    <svg
      fill="currentColor"
      height="16px"
      viewBox="0 0 14.01 14"
      width="16px"
      xmlns="http://www.w3.org/2000/svg"
    >
      <defs />
      <g>
        <g>
          <path
            d="M10.5 7h-1a.5.5 0 010-1h1a.5.5 0 010 1zm-3 0h-1a.5.5 0 010-1h1a.5.5 0 010 1zm-3 0h-1a.5.5 0 010-1h1a.5.5 0 010 1zm6 3h-1a.5.5 0 010-1h1a.5.5 0 010 1zm-3 0h-1a.5.5 0 010-1h1a.5.5 0 010 1zm-3 0h-1a.5.5 0 010-1h1a.5.5 0 010 1zm-3-6.95a.53.53 0 00-.5.53v7.93a.54.54 0 00.53.54h10.94a.54.54 0 00.53-.54V3.58a.53.53 0 00-.53-.53h-1.33a.13.13 0 00-.1 0 .12.12 0 000 .1v.35a.5.5 0 01-1 0v-.31a.14.14 0 00-.14-.14H4.14a.14.14 0 00-.14.14v.31a.5.5 0 01-1 0v-.31a.12.12 0 000-.1.13.13 0 00-.1 0zM9.86 2a.14.14 0 00.14-.14V1.5a.5.5 0 011 0v.36a.14.14 0 00.14.14H13a1 1 0 011 1v9a1 1 0 01-1 1H1a1 1 0 01-1-1V3a1 1 0 011-1h1.86A.14.14 0 003 1.86V1.5a.5.5 0 011 0v.36a.14.14 0 00.14.14z"
            fill="currentColor"
            fill-rule="evenodd"
            id="calendar-outlined_svg__calendar"
          />
        </g>
      </g>
    </svg>
  </span>
  <span>
    今天
  </span>
</button>
`;
